frontend/components/maps/CensusTractMap.js

```diff
@@ -0,0 +1,179 @@
+        layer.bindPopup(popupContent);
+    }
+
+    /**
+     * Generates the styling for a given census tract data point, and returns the style object
+     * required by the Leaflet GeoJSON component
+     *
+     * @param tract_data element of state.census_tracts, the tract to determine styling for
+     * @returns {{color: (string)}} object to be used by GeoJSON.style
+     */
+    styleTracts(tract_data) {
+        let tract_name = tract_data.properties["GISJOIN"];
+
+        // Deanwood is itself, so color that specially
+        if (tract_name === DEANWOOD_JOIN_CODE) return {color: "black", fillColor: "#00FF00"};
+
+        // Color tract different if there is no data for it
+        if (!(tract_name in this.props.deanwood_similarities)) return {color: "grey"};
+
+        const tract_distance = this.props.deanwood_similarities[tract_name]["distance"];
+
+        return {
+            // color: this.calculateTractColor(tract_distance)
+            color: "black",
+            fillColor: "#00FF00",
+            fillOpacity: 0.8 * (this.state.farthest_distance/5 - tract_distance) /
+                (this.state.farthest_distance/5 - this.state.closest_distance),
+            opacity: 0.8 * (this.state.farthest_distance/5 - tract_distance) /
+                (this.state.farthest_distance/5 - this.state.closest_distance)
+        };
+    }
+
+    prep_distance_data() {
+        Object.entries(this.props.deanwood_similarities).forEach(([_, data]) => {
+            const distance = data["distance"];
+            this.setState({
+                closest_distance: Math.min(this.state.closest_distance, distance),
+                farthest_distance: Math.max(this.state.farthest_distance, distance)
+            });
+        });
+        return true;
+    }
+
+    render() {
+        return (
+            <div id="map">
+                <FormSelect
+                    aria-label="Census Map Focus Selector"
+                    onChange={(e) => this.onPosSelected(e)}
+                >
+                    <option value="deanwood">Deanwood</option>
+                    <option value="detroit">Detroit</option>
+                    <option value="los_angeles">Los Angeles</option>
+                    <option value="houston">Houston</option>
+                </FormSelect>
+                <MapContainer
+                    center={this.state.start_pos}
+                    zoom={11} scrollWheelZoom={true}
+                    whenCreated={map => this.setState({map})}
+                >
+                    <TileLayer
+                        attribution='&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors'
+                        url="https://stamen-tiles-a.a.ssl.fastly.net/toner-lite/{z}/{x}/{y}.png"
+                    />
+                    {this.props.census_tracts && this.props.deanwood_similarities &&
+                        this.props.census_tracts.map(tract => {
+                            return (
```

Comment:
@anastasia mentioned a performance problem on this page, and it stems from here. Right now, on each render, `styleTracts` and `onEachFeature` is getting called for each of ~7500 tracts, and that's causing the page to be quite laggy.

You could try a few things...

Something like:
```
    shouldComponentUpdate(nextProps, _nextState, _nextContext) {
        return (nextProps !== this.props);
    }
```
will get rid of a bunch of unnecessary renders. e.g., switching the city selector right now causes an unnecessary rerender.

That doesn't get rid of an initial relatively slow loop through all of the tracts. Some thoughts there:
- it looks like there's a bunch of work happening in the frontend here that could be precomputed on the backend
- can you defer adding a GeoJSON to the MapContainer until it's actually in the viewport?

Reply:
I tinkered with it a bit, and I think I got the performance back to something reasonable. I couldn't find a way to prevent adding the GeoJSON objects until they were in the viewport (at least, not without querying the backend), but I was able to make a lot of the computation done in advance and just stored in the json object that gets sent to the front end. I also ended up cutting the number of tracts that the frontend deals with about in half (basically I made it such that the GeoJSON feature isn't even sent to the frontend if it isn't within a certain threshold). Hopefully all of this ends up helping, but please let me know! It's a lot more fluid on my computer, but I'm not sure how it'll hold up on others.